### PR TITLE
Format smart now accepts string as amount

### PR DESCRIPTION
**Pull request by alfetopito**
# Update 2020-11-10

Exposing `stringToBn` function.
Plus addressing comments.

---

# Summary

On `format*` utils functions, I'd like to pass a number as a string and receive the properly formatted number as string back, without having to convert it to a BN first.
I may or may not need additional precision adjustment.

## Example:

``` typescript
formatSmart('0.000001', 0) // => '< 0.001'
``` 

## Background

We originally developed these helper functions to deal with BNs and wei/unit conversions.
On CMM app we are using [Decimal js](https://mikemcl.github.io/decimal.js) and if I want to use these nice functions I need to go a long way converting values from Decimal -> string -> BN, which is a bit annoying as this implementation shows. (Unless, there's a better way?)

See for instance a wrapper I created around `formatSmart` -> https://github.com/gnosis/safe-cmm-app-react/blob/b951c76f580358b9dbe507140c63fdfdaf2abb9f/src/utils/format.ts

So I'm proposing making all functions accept the amount as string, to make it number library agnostic.
I'm not going to remove BN from the input for backwards compatibility, but maybe we should consider a breaking change in the future to unify the BN|BigNumber|Decimal mess we have =P

### CODE REVIEW

File: src/utils/format.ts [hunk 1]
```diff
@@ -132,6 +133,38 @@ function _decomposeBn(amount: BN, amountPrecision: number, decimals: number): {
...
+ * We do so by creating a BigNumber and moving the decimal separator to the right
+ * by multiplying it by 10^(amount precision).
+ * Which we then later add to regular precision to be 'compressed'.
+ * Finally, convert it to a string base 10,
+ * to avoid numbers with exponents which BN also doesn't like
+ *
+ * @param amountStr Amount with arbitrary precision
+ * @param amountPrecision precision on top of existing precision
+ */
+function _stringToBn(amountStr:string, amountPrecision:number):{ amount: BN, precision: number } {
```
Comment (anxolin) [inline src/utils/format.ts]: This function is very useful and general purpose, I think it should be exposed. In fact, @W3stside have been lately doing some conversions lately between BigNumber, String and BN, I think we should have this tools available. I think this PR might be useful for him too.

For example, for other PRs might be interesting:
* `stringToBn`: this one
* `bnToBigNumber`: BN+Precission --> BigNumber
* `bigNumberToBN`: BigNumber+Precission --> BN
* `stringToBigNumber`: amount+precission-->BigNumber

ubernit: what about simplifying the param names?
```ts
amount:string, precision:number
```

File: src/utils/format.ts [hunk 2]
```diff
@@ -132,6 +133,38 @@ function _decomposeBn(amount: BN, amountPrecision: number, decimals: number): {
...
+  const bigNumberAmount = new BigNumber(amountStr)
+
+  const originalPrecision = bigNumberAmount.precision()
```
Comment (anxolin) [inline src/utils/format.ts]: Isn't it `bigNumberAmount.precision(true)`

image

File: src/utils/format.ts [hunk 3]
```diff
@@ -132,6 +133,38 @@ function _decomposeBn(amount: BN, amountPrecision: number, decimals: number): {
...
+
+  const amount = new BN(bigNumberAmount
+    .multipliedBy(TEN_BIG_NUMBER
+      .pow(originalPrecision))
+    .integerValue()
+    .toString(10))
+
+  const precision = originalPrecision + (amountPrecision as number)
```
Comment (anxolin) [inline src/utils/format.ts]: Why does it need casted if it was already declared as number?

Comment (anxolin) [inline src/utils/format.ts]: Why the resulting `precision` is the sum of the original and the `amountPrecision`, also (and related), why do we want to return a "new precision"?

Maybe wrong, but my thinking is, I want to format something with precision 4. So no matter what, the precission is always 4.
`amount=12.20, precision=4` --> `BN=122000`

File: test/utils/format/formatSmart.spec.ts [hunk 4]
```diff
@@ -180,3 +180,26 @@ describe('Edge cases', () => {
...
+
+describe('Amount is a string', () => {
+  test('12,345.67 - Precision 0 (just adjust formatting)', () => {
+    const amount = '12345.67'
+    expect(formatSmart(amount, 0)).toEqual('12,345.67')
+  })
+  test('12,345.6789 - More decimals than precision', () => {
+    const amount = '1234567.89'
+    expect(formatSmart({ amount, precision: 2, decimals: 4 })).toEqual('12,345.6789')
+  })
```
Comment (anxolin) [inline test/utils/format/formatSmart.spec.ts:193]: I might be wrong, but maybe a test that has 0s left of the dot might fail? '1234500.89'

Also, nice to add some extra test with amount without decimals, or with more decimals than the precision. I feel some edge cases are still missing.

**Review by anxolin — COMMENTED** [review]
Good call, 👌
this is super useful

File: src/utils/format.ts (hunk 1 above)
Comment (alfetopito) [reply src/utils/format.ts]: Makes sense. Will expose it.

As for the parameter names, the reason for it is that `amount` and `precision` are exactly what I'm returning.
So I changed to avoid confusion between input/output.

File: src/utils/format.ts (hunk 2 above)
Comment (alfetopito) [reply src/utils/format.ts]: My understanding was that, if I do provide `true`, it'll return precision including trailing zeros, which I do not case about.
Or was that on Decimal.js? Damn, too similar yet slightly different interfaces.

No yeah, that's exactly it -> https://mikemcl.github.io/bignumber.js/#sd

I don't want to keep trailing zeros:

```typescript 
new BigNumber('1.2300').precision(true) \\ => 4
new BigNumber('1.2300').precision() \\ => 2
```

Did not actually test with `true`, though

File: src/utils/format.ts (hunk 3 above)
Comment (alfetopito) [reply src/utils/format.ts]: Left over from when it was inside `formatSmart`.
Will remove

Comment (alfetopito) [reply src/utils/format.ts]: Because I might want to keep the original precision.

I said in the description that I want to preserve the original precision and maybe add on top of it.

The use case is to deal with numbers that are decimals in the correct precision already, and I just need to properly format them.

With a further refactor of how we do the formatting this might be skipped entirely, but since right now we work with BNs, I need to add the original precision so it can be removed later and give me back the original value.

Example:
I want to nicely format the number `1234.56789`.
I expect to receive a string with decimal separator and with the expected amount of decimals, such as `1,234.5679`
There's no need to adjust the precision here. So I'll call `formatSmart('1234.56789', 0)`.

Internally (or externally like I'm currently doing) I need to first convert this string representation of a decimal number to BN.
Thus, I return the original precision together with the amount as BN, so I know how to format it back to it's original form.

That would result in `formatAmount(BN('123456789'), 5)`

Is that more clear?

File: test/utils/format/formatSmart.spec.ts (hunk 4 above)
Comment (alfetopito) [reply test/utils/format/formatSmart.spec.ts:193]: I'll add more tests :)

File: src/utils/format.ts (hunk 2 above)
Comment (alfetopito) [reply src/utils/format.ts]: Actually, what I'm looking for is `decimalPlaces()`, not `precision`. https://mikemcl.github.io/bignumber.js/#dp :facepalm:

Comment (anxolin) [reply src/utils/format.ts]: cool 👌

File: src/utils/format.ts (hunk 3 above)
Comment (anxolin) [reply src/utils/format.ts]: More or less. I think it's clear, I also notice now the JSDoc in the param `amountPrecision precision on top of existing precision`

I find it a bit weird the signature of this, but nvm, right now I cannot put the finger on what is exactly

Comment (anxolin) [reply src/utils/format.ts]: If it behaves like that, should we make it more clear in the param then: `precissionToAdd` so it's more clear.
seeing `_stringToBn(amount='123.45', precission=0)` I was initially expecting to return sth like `123` (so u know the precision is 0)

...or `_stringToBn(amount='123.4567', precission=5)` returning 12345670

But I think u wanted to provide different utility to this function, and u really want to add up the precisions, maybe is a matter of changing the name to `amountPrecission` param?

**Review by anxolin — APPROVED** [review]

**Comment by anxolin** [comment]
@alfetopito should we merge this?
